var/spack/repos/builtin/packages/gdal/package.py

```diff
@@ -128,6 +129,7 @@ class Gdal(AutotoolsPackage):
     depends_on('fyba', when='+sosi')
     depends_on('hdf', when='+hdf4')
     depends_on('hdf5', when='+hdf5')
+    depends_on('hdf5@:1.12', when='@:3.4.1 +hdf5')
```

Comment:
This may be overly restrictive -- it might be able to compile with newer versions with variant `api=112`.

Reply:
Ooh, interesting. I think it might be difficult to properly specify such a dep. I guess we could do something like:
```python
depends_on('hdf5', when='+hdf5')
conflicts('^hdf5@1.13: api=default')
conflicts('^hdf5@1.13: api=v114')
```
but that will be out of date as soon as a new api is added to the list. Let's keep it as is for now until someone actually needs to build old GDAL with new HDF5.